src/persist/README.md

```diff
@@ -3,49 +3,194 @@
+writes by this process (they can be opened for reads by other processes,
+however).
 
-[persister v1]: https://github.com/danhhz/differential-dataflow/blob/02673114b05933341893ab603327237a9583e432/persist/src/persister.rs#L30-L39
+The runtime stops when every copy of the client has been dropped (or when any
+client's stop method has been called), attempting to first gracefully finish any
+pending work, but rejecting new work.
 
-```sql
-CREATE TABLE persisted (persisted_id INT, key BYTES, val BYTES, ts BYTES, diff BYTES);
-CREATE INDEX ON persisted (persisted_id, ts, key);
+As needed, the user creates a persisted stream by name, returning a token. At
+dataflow construction time, this token can be exchanged to create one of several
+available persistence operators (see below). When the operator is run, it hands
+any blocking io or cpu heavy work to the runtime, asking the dataflow to be
+scheduled again when this finishes. If the same buffer and blob implementations
+were previously used to persist a stream with the same name, this previous data
+is read and also output by the operator, yielding periodically to allow
+downstream operators a chance to process it.
+
+## Operator Guarantees
+
+- _idempotence_: Any record emitted by a persistence operator is guaranteed to
+  be included in the output of all future operators for the same persistent
+  stream. This is true even before the record's timestamp has been closed.
+- _ordering_: The same guarantees you'd expect out of any differential
+  collection or timely stream. TODO: Is there anything more specific to say
+  here?
+- _determinism_: TODO: This is easy in the absence of errors from the storage,
+  but that's not the interesting part. "S3 is down" is obviously a transient
+  error, but even things like permissions errors are transient in terms of the
+  strict definition of deterministic we're using. I think the Materialize "error
+  stream" model is potentially a good way to model this where anything on the
+  main output is deterministic at some timestamp iff the error stream is an
+  empty collection at that timestamp. Think about this more.
+
+## Example: Persisting a Stream
+
+NB: This operator doesn't exist yet, but this is more or less what the interface
+will look like.
+
+```rust
+let persister: RuntimeClient = ...
+let token = persister.create_or_load("example stream")?;
+
+let stream: Stream<(K, V, T, R)> = ...
+let (persisted_stream, err_stream) = stream.new_persisted_stream(token);
+```
+
+Whenever the operator is woken up to do work, it sends off any new inputs to be
+persisted. Any previous inputs that have finished being durably written are
+forwarded on as an operator output. When the input frontier advances, the output
+fronter is advanced to match, but only after all relevant input data has been
+persisted.
+
+Any errors encountered while persisting (storage unavailable, etc) are emitted
+to an error stream.
```

Comment:
If the errors are not durably persisted, and are instead reflections of the persistence layer's current behavior, they are going to be a new source of non-determinism in `dataflow` which is potentially problematic. At least, it would be clearest if the persistence layer only produces data known to be correct, and we would need to change the semantics if it is going to be "not the correct answer, but instead info about why we don't have the correct answer".

E.g. if I take two of these streams and subtract them, do I get the always-empty stream, or a stream that has some positive errors in it and some negative errors in it, as a function of when in the replaying of the persisted streams there was a S3 hiccup?

Reply:
Ignore what I have here, like the idempotence stuff it was just a guess at what makes the most sense and I'm not tied to it. From my perspective, everything we can do here is equivalently easy.

"it would be clearest if the persistence layer only produces data known to be correct" sounds to me like we just don't emit anything while encountering a non-deterministic error and just hold up the capability. Is that right?

Reply:
That's what I would start with, personally. We could dive a bit deeper on what errors "should mean" and perhaps we end up relaxing guarantees for errors (i.e. not always a deterministic function of the input) but we should probably do so intentionally. It may totally be that we are already naughty in some other sources, and I missed it though. But imo, yeah: if the right data are known absent at the moment, best not to say anything into the dataflow at least.

Reply:
Removed everything about errors in the new version until someone gets a chance to figure out how they should work. I did note that any solution for them should result in definite inputs to dataflow, which I think concisely covers this concern (and others)